internal/overlord/logstate/manager_test.go

```diff
@@ -0,0 +1,238 @@
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+	. "gopkg.in/check.v1"
+)
+
+type managerSuite struct {
+	logbuf        *bytes.Buffer
+	restoreLogger func()
+}
+
+var _ = Suite(&managerSuite{})
+
+func Test(t *testing.T) { TestingT(t) }
+
+func (s *managerSuite) SetUpTest(c *C) {
+	s.logbuf, s.restoreLogger = logger.MockLogger("PREFIX: ")
+}
+
+func (s *managerSuite) TearDownTest(c *C) {
+	s.restoreLogger()
+}
+
+func (s *managerSuite) TestSelectTargets(c *C) {
+	unset := plan.LogTarget{Selection: plan.UnsetSelection}
+	optout := plan.LogTarget{Selection: plan.OptOutSelection}
+	optin := plan.LogTarget{Selection: plan.OptInSelection}
+	disabled := plan.LogTarget{Selection: plan.DisabledSelection}
+
+	input := plan.Plan{
+		LogTargets: map[string]*plan.LogTarget{
+			"unset":    &unset,
+			"optout":   &optout,
+			"optin":    &optin,
+			"disabled": &disabled,
+		},
+		Services: map[string]*plan.Service{
+			"svc1": {LogTargets: nil},
+			"svc2": {LogTargets: []string{}},
+			"svc3": {LogTargets: []string{"unset"}},
+			"svc4": {LogTargets: []string{"optout"}},
+			"svc5": {LogTargets: []string{"optin"}},
+			"svc6": {LogTargets: []string{"disabled"}},
+			"svc7": {LogTargets: []string{"unset", "optin", "disabled"}},
+		},
+	}
+
+	expected := map[string]map[string]*plan.LogTarget{
+		"svc1": {"unset": &unset, "optout": &optout},
+		"svc2": {"unset": &unset, "optout": &optout},
+		"svc3": {"unset": &unset},
+		"svc4": {"optout": &optout},
+		"svc5": {"optin": &optin},
+		"svc6": {},
+		"svc7": {"unset": &unset, "optin": &optin},
+	}
+
+	planTargets := selectTargets(&input)
+	c.Check(planTargets, DeepEquals, expected)
+	// Check no error messages were logged
+	c.Check(s.logbuf.Bytes(), HasLen, 0)
+}
+
+func (s *managerSuite) TestLogManager(c *C) {
+	m := NewLogManager()
+	// Fake ringbuffer so that log manager can create forwarders
+	rb := servicelog.RingBuffer{}
+
+	// Call PlanChanged with new plan
+	m.PlanChanged(&plan.Plan{
+		Services: map[string]*plan.Service{
+			"svc1": {},
+			"svc2": {LogTargets: []string{"optin", "disabled"}},
+			"svc3": {LogTargets: []string{"unset"}},
+		},
+		LogTargets: map[string]*plan.LogTarget{
+			"unset":    {Name: "unset", Type: plan.LokiTarget, Selection: plan.UnsetSelection},
+			"optout":   {Name: "optout", Type: plan.LokiTarget, Selection: plan.OptOutSelection},
+			"optin":    {Name: "optin", Type: plan.LokiTarget, Selection: plan.OptInSelection},
+			"disabled": {Name: "disabled", Type: plan.LokiTarget, Selection: plan.DisabledSelection},
+		},
+	})
+
+	// Start 3 services
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	// Call ServiceStarted
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc1", &rb)
+	}()
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc2", &rb)
+	}()
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc3", &rb)
+	}()
```

Comment:
Can't we just do the following inline / synchronously, because `ServiceStarted` just creates an object an fires up the `forward` goroutine, right, so it'll be super-quick?

```
m.ServiceStarted("svc1", &rb)
m.ServiceStarted("svc2", &rb)
m.ServiceStarted("svc3", &rb)
```

Reply:
I was trying to make this "realistic" in that Pebble can start services concurrently (I assume?). Happy to make it synchronous if you think this doesn't matter for the test.

Reply:
True, not a bad idea (and `-race` *might* pick up any issues if we're lucky). In that case maybe just a comment saying why you're doing it this way.